### PR TITLE
DEV: Refactor DM channel creation into new service pattern

### PR DESCRIPTION
**Pull request by martin-brennan**
This will be used when we move the channel creation for DMs
to happen when we first send a message in a DM channel to avoid
a double-request. For now we can just have a new API endpoint
for creating this that the existing frontend code can use,
that uses the new service pattern.

This also uses the new policy pattern for services where the policy
can be defined in a class so a more dynamic reason for the policy
failing can be sent to the controller.

### CODE REVIEW

File: plugins/chat/app/helpers/chat/with_service_helper.rb
Comment (martin-brennan) [inline plugins/chat/app/helpers/chat/with_service_helper.rb:23]: LMK your thoughts on the changes here @Flink , I did them so we can also use `with_service` in specs, e.g. see https://github.com/discourse/discourse/pull/22144/files#diff-fffd7577aaa7e572a63261269cd5b0a1ada0b80b744799412d110a717f50c376R21

**Comment by martin-brennan** [comment]
@jjaffeux @Flink I closed off https://github.com/discourse/discourse/pull/21216 rather than trying to fix merge conflicts. Loic, can you please take a look and LMK what you think, especially of the couple of changes I have made to `with_service` to use in specs. I also made one big policy for the `user_comm_screener` stuff like we discussed in the other PR. Hopefully we can get this merged soon, sorry I spent so long getting the old one finished after Loic merged in the policy improvements.

File: plugins/chat/spec/services/chat/create_direct_message_channel_spec.rb
Comment (Flink) [inline plugins/chat/spec/services/chat/create_direct_message_channel_spec.rb:84]: `fab!` and `let` should appear before `before` blocks.

File: plugins/chat/app/helpers/chat/with_service_helper.rb
Comment (martin-brennan) [reply plugins/chat/app/helpers/chat/with_service_helper.rb:23]: @Flink yes, @jjaffeux and I have discussed this and a few weeks back we changed the `:chat_message` fabricator to use the `Chat::MessageCreator` service, since we want things to reflect reality more especially for system specs. We will need to do something like this if we ever want to call services outside of controllers too...

Joffrey what are your thoughts on this? I don't have a super strong opinion on it so happy to defer to you/Loic

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (martin-brennan) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:45]: I think it makes sense, since otherwise we would have to introduce logic back into the controller level (which is calling the service) to do the same thing. IMO better to just take whatever `params[:target_usernames]` is and sort it out in the contract before validation.

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (martin-brennan) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:62]: Cool sounds good :+1:

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (martin-brennan) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:96]: It appears so! Changed it (also I had to use `upsert_all`)

File: plugins/chat/spec/services/chat/create_direct_message_channel_spec.rb
Comment (martin-brennan) [reply plugins/chat/spec/services/chat/create_direct_message_channel_spec.rb:84]: I wish we could have rubocop rules for this and the subject one

**Comment by martin-brennan** [comment]
@Flink did first round of review fixes + comments :)

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (jjaffeux) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:45]: > I think it makes sense, since otherwise we would have to introduce logic back into the controller level (which is calling the service) to do the same thing. IMO better to just take whatever `params[:target_usernames]` is and sort it out in the contract before validation.

Not sure I follow. Can't you just handle differently the `target_usernames` param inside the service based on its type?

File: plugins/chat/app/controllers/chat/api/direct_messages_controller.rb
Comment (jjaffeux) [inline plugins/chat/app/controllers/chat/api/direct_messages_controller.rb:25]: I think we try to be too verbose here.

I suggest the following:

```
on_failed_policy(:satisfies_max_direct_message_users)
on_failed_policy(:targets_allow_dms)
on_failed_policy(:targets_allow_dms_from_user)
```

File: plugins/chat/app/controllers/chat/api/direct_messages_controller.rb
Comment (jjaffeux) [inline plugins/chat/app/controllers/chat/api/direct_messages_controller.rb:1]: I wonder if its worth creating all this code to delete it soon?

File: plugins/chat/app/helpers/chat/with_service_helper.rb
Comment (Flink) [reply plugins/chat/app/helpers/chat/with_service_helper.rb:23]: Well, I guess it’s always about trade-offs :sweat_smile: Indeed, by using the service, we’re sure to have the exact same output than in production. But as you said @jjaffeux, testing the side effects should happen in the service specs and not outside. In other specs, if we need to create a DM channel, then it’s probably just a setup step and I’m pretty sure most of the time we won’t be interested by anything else than just having a channel properly created.

Using fabricators only, we have to replicate some of the logic we have in the service and have to keep it up-to-date. In the end, it depends on what we need. I think using fabricators should be enough but I could be wrong :slightly_smiling_face: 

If we wanted to use the service directly, actually we could do this by using fabricators too as they’re able to call any object.

Finally, we don’t really need `with_service` in specs, we can simply call the service directly and provide proper arguments as we control everything here. We shouldn’t have to handle the service failures as we want a specific state for our tests, so if it’s failing then it means our setup is not a good one.

File: plugins/chat/spec/services/chat/create_direct_message_channel_spec.rb
Comment (Flink) [reply plugins/chat/spec/services/chat/create_direct_message_channel_spec.rb:82]: Going to check if they already exists :slightly_smiling_face:

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (Flink) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:45]: This is what the contract does at the moment, no? If it’s a string, split it into an array, otherwise don’t do anything.

Comment (jjaffeux) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:45]: Sorry that wasn't clear, I meant we could have two params maybe?

Comment (Flink) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:45]: By the way, I’ve found a way of simplifying the contract a bit:
```ruby
    class Contract
      attribute :target_usernames

      before_validation { self.target_usernames = target_usernames.split(",").flatten }

      validates :target_usernames, presence: true
    end
```
`#split` is defined both on `Array` and `String` so we can just call it without having to perform a check first. Then `#flatten` is useful because `Array#split` will return an array of arrays.

I’ve removed `length: { minimum: 1 }` because we’re already validating presence, so the validation will pass only if there is at least one element.

Comment (Flink) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:45]: > Sorry that wasn't clear, I meant we could have two params maybe?

I think it would complexify things instead. Ideally, if we could do something like `attribute :target_usernames, :array` with auto-casting, it would be perfect :grin:

File: plugins/chat/spec/services/chat/create_direct_message_channel_spec.rb
Comment (Flink) [reply plugins/chat/spec/services/chat/create_direct_message_channel_spec.rb:82]: I haven’t found an existing rule for `let` and `before` blocks, but there is one for `subject`: https://docs.rubocop.org/rubocop-rspec/cops_rspec.html#rspecleadingsubject
I’m going to open a PR against our rubocop gem and see if this is something we could merge quickly.

Comment (Flink) [reply plugins/chat/spec/services/chat/create_direct_message_channel_spec.rb:82]: I’ve opened two PRs:
https://github.com/discourse/rubocop-discourse/pull/32
https://github.com/discourse/discourse/pull/22225

File: plugins/chat/app/controllers/chat/api/direct_messages_controller.rb
Comment (martin-brennan) [reply plugins/chat/app/controllers/chat/api/direct_messages_controller.rb:1]: I mean, it's just the `with_service` stuff in the controller, the actual service code won't be removed, so IMO it's fine for now.

File: plugins/chat/app/controllers/chat/api/direct_messages_controller.rb
Comment (martin-brennan) [reply plugins/chat/app/controllers/chat/api/direct_messages_controller.rb:25]: These make sense except the second one, it should be more like `actor_allows_dms`. Will change :+1:

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (martin-brennan) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:45]: Thanks @Flink will go with the simplified contract. I too think adding 2 params will be needlessly complex when we can just accept both in 1 param and fix it in the contract.

File: plugins/chat/app/helpers/chat/with_service_helper.rb
Comment (martin-brennan) [reply plugins/chat/app/helpers/chat/with_service_helper.rb:23]: > If we wanted to use the service directly, actually we could do this by using fabricators too as they’re able to call any object.

Yeah this is what we are already doing with the `:chat_message` fabricator.

> We shouldn’t have to handle the service failures as we want a specific state for our tests, so if it’s failing then it means our setup is not a good one.

I suppose this is a good point, I mainly wanted a nice way to automatically get a spec failure + be able to inspect it when a called service fails rather than having to do it manually in each test.

> Maybe using it for Chat::Message was not a good idea too, and the fact that we thought it was is probably showing us shortcomings...

It is hard to tell...the side-effects are really necessary for system specs IMO for them to be realistic, but yeah probably overkill for normal unit tests like this :thinking: Maybe we could have a version of the fabricator which uses the service class and a version without and use the appropriate one based on the test? E.g. `Fabricate(:chat_message_with_service)` for system tests and `Fabricate(:chat_message)` for unit specs? Not sure...

@jjaffeux @Flink I have no idea how github notifies based on comments so just mention you both again :)

Comment (Flink) [reply plugins/chat/app/helpers/chat/with_service_helper.rb:23]: Indeed, if we need message bus events to be sent or similar things, it’s better to call the service directly. This idea of having different fabricators also crossed my mind, so maybe that’s something to explore?

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (Flink) [inline plugins/chat/app/services/chat/create_direct_message_channel.rb:85]: We can probably remove this comment :slightly_smiling_face:

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (Flink) [inline plugins/chat/app/services/chat/create_direct_message_channel.rb:101]: There’s a shortcut to write this, `Time.current` :grin:

**Comment by jjaffeux** [comment]
We have been sitting on this for long time for various reasons :p Let's try to merge it this week. I let you settle on the last points and will try it locally a lot wednesday

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (Flink) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:45]: I’ve just created this small PR: https://github.com/discourse/discourse/pull/22278.
Having a simple way of handling arrays is always nice, so tell me what you think :grin: @jjaffeux @martin-brennan

Comment (martin-brennan) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:45]: Thanks @Flink I merged that PR and will use it here :+1:

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (martin-brennan) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:101]: Didn't know that, but we use `Time.zone.now` way more often in the codebase so I would rather it be consistent

**Comment by martin-brennan** [comment]
Okay @jjaffeux and @Flink have made those last few fixes, let's get this merged tomorrow if possible :pray:

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (jjaffeux) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:83]: Won't die on this hill, but really I dont think using `Time.current` could be classified as being inconsistent. It's also something you can grep/replace in a commit if that is an issue.

File: plugins/chat/app/helpers/chat/with_service_helper.rb
Comment (jjaffeux) [reply plugins/chat/app/helpers/chat/with_service_helper.rb:23]: Yep Im not sure yet too... so far having fabricators without side effects have caused me way more issues than having fabricators being the actual service call, that's all I can tell.

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (Flink) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:83]: I think at some point Rubocop was suggesting to use `Time.current` but it seems both forms are equally accepted (`Time.current` is just shorter). That’s basically what is said in the Rails style guide: https://rails.rubystyle.guide/#time-now :sweat_smile:

File: plugins/chat/app/helpers/chat/with_service_helper.rb
Comment (Flink) [reply plugins/chat/app/helpers/chat/with_service_helper.rb:23]: Indeed, when you need all the “real” effects like in system specs it’s very convenient. But in all the other specs, I think it’s overkill (and slower, etc.)

Comment (Flink) [reply plugins/chat/app/helpers/chat/with_service_helper.rb:23]: So, are we keeping this in the end? I’m pretty convinced now that having `with_service` in specs is not necessary, but we can just call the service in fabricators (this way it’s relatively consistent with how we setup things in specs).

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (Flink) [inline plugins/chat/app/services/chat/create_direct_message_channel.rb:41]: `length` is not necessary, as we saw previously :slightly_smiling_face:

**Review by Flink — APPROVED** [review]
Just some code to remove I think but otherwise I think we’re pretty good :grin:

File: plugins/chat/app/helpers/chat/with_service_helper.rb
Comment (martin-brennan) [reply plugins/chat/app/helpers/chat/with_service_helper.rb:23]: TBH I don't want to have to mess around with changing the fabricator for DM channel at the moment, I want to get this merged, so I will just revert the `with_service` changes in this PR and call the service directly in the specs

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (martin-brennan) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:83]: `Time.current` has 16 results in core, `Time.zone.now` has 361, so I would class that as inconsistent :) Leaving it as is

**Comment by martin-brennan** [comment]
@jjaffeux last changes have been made, LMK if there are any issues when you try it out today, will leave it unmerged for now

**Review by jjaffeux — APPROVED** [review]
Nice work @Flink and @martin-brennan thanks

Let's try to merge early in your day when you can martin

File: plugins/chat/app/services/chat/create_direct_message_channel.rb
Comment (jjaffeux) [inline plugins/chat/app/services/chat/create_direct_message_channel.rb:95]: Can't we concat an object into memberships for the guardian user and do it all in one upsert?

Comment (martin-brennan) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb:95]: Hmm actually I have a feeling we may already do everyone in the first upsert, will try just removing this line

Comment (jjaffeux) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb]: I would say so yes:

```ruby
def fetch_target_users(guardian:, contract:, **)
  User.where(username: [guardian.user.username, *contract.target_usernames]).to_a
end
```

Comment (martin-brennan) [reply plugins/chat/app/services/chat/create_direct_message_channel.rb]: Yep I checked locally and removed it :+1:

**Pull request merged**
